Migrating to Asciidoc

maxim-tschumak opened the pull request:
This PR implements the migration from Markdown/GitBook to Asciidoc(tor) and includes:

* Translation from Markdown to Asciidoc markup
* New build and deployment scripts
* Introduces RuleId concept and concrete rule ids (and tooling around them)
* Support for the existing links (redirects to the current rule locations)
* Adjusted documentation
* Updated Maintainers list

Please also have a look at the [generated HTML](https://maxim-tschumak.github.io/restful-api-guidelines/).

@tfrauenstein Feel free to award the rule ids [1,99] to the most prominent rules.

Related to #231, #69, #10, #238.
Closes #260.

duergner commented:
@maxim-tschumak what's the reason to move over to *RuleId* instead of having speaking URLs?

maxim-tschumak commented:
@duergner It's a trade-off: having _short easy-to-remember_ RuleIds or _human-readable_ links. As discussed in the Guild, we prefer short identifiers.
It also makes rule identifier reliable (doesn't change when the rule title changes) and allows integrations of another tooling like [Zally](https://github.com/zalando-incubator/zally).

ePaul commented on MAINTAINERS:
```diff
@@ -1,5 +1,14 @@
+Tim Lossen <tim.lossen@zalando.de>
+Andreas Schmidt <andreas.schmidt@zalando.de>
 Willi Schönborn <willi.schoenborn@zalando.de>
+Maxim Tschumak <maxim.tschumak@zalando.de>
```
Many new maintainers ... Welcome everyone ;-)

ePaul commented on README.adoc:
```diff
@@ -0,0 +1,31 @@
+to challenge their APIs. As such, you should consider this to be a
+living, evolving document. We will revise and update based on our
+learnings and experiences.
+
```
Maybe put a link to the BUILD file here?

ePaul commented on SUMMARY.adoc:
```diff
@@ -0,0 +1,23 @@
+= Summary
+
+* link:chapters/design-principles.adoc[Design Principles]
```
Just for understanding, how does the introduction appear in the output? It is not listed here.

ePaul replied:
Ah, it comes from index.adoc. Why do we have this file here, then?

maxim-tschumak replied:
I removed `index.adoc` -> I also see no need for it.

ePaul commented:
Just a general hint for the next time:
Some files totally change the line breaks, without any content change.
This makes it much harder to review that actually not much changed, as Github's diff view doesn't even notice that they are in fact the same file.

ePaul commented on chapters/common-data-types.adoc:
As this is an internal link, maybe mark is as such?

ePaul commented on chapters/compatibility.adoc:
eficits → deficits.
(I know, this didn't change really with this PR, but as everything is marked as changed anyways ...)

ePaul commented on chapters/compatibility.adoc:
objects → object

ePaul commented on chapters/data-formats.adoc:
```diff
@@ -0,0 +1,112 @@
+[#168]
+== {MAY} Use non JSON Media Types for Binary Data or Alternative Content Representations
+
+Other media types may be used in following cases: * Transferring binary
```
Here some line breaks (and therefore the list structure) got lost.

ePaul commented on chapters/data-formats.adoc:
```diff
@@ -0,0 +1,112 @@
+
+=== HTTP headers
+
+Http headers including the proprietary headers. Use the
```
Not new, but this is not a sentence. Can we get rid of the `.` and just join the two sentences here?

ePaul commented on chapters/data-formats.adoc:
```diff
@@ -0,0 +1,112 @@
+|=====================================================================
+|type |format |specified value range
+|integer |int32 |integer between -231 and 231-1
+|integer |int64 |integer between -263 and 263-1
```
Those ranges are a bit small ;-)

The exponents got lost here.

ePaul commented on chapters/design-principles.adoc:
```diff
@@ -0,0 +1,94 @@
+
+As mentioned above, Zalando is transforming from an online shop into an
+expansive fashion platform comprising a rich set of products following a
+Software as a Platform (SaaS) model for our business partners. As a
```
Hmm, "Software as a Platform" doesn't have the abbreviation "SaaS" (that is software as a service).

@tfrauenstein Is the abbreviation or the full name wrong? (It was added in your PR #159.)

maxim-tschumak replied:
I've corrected it to `SaaP`.

ePaul commented on chapters/design-principles.adoc:
```diff
@@ -0,0 +1,94 @@
+Law, aka http://en.wikipedia.org/wiki/Robustness_principle[the
+Robustness Principle] (RFC 1122):
+
+* Be liberal in what you accept, be conservative in what you send
```
Hmm, not sure if that is still applicable, as we derive from it in the Compatibility chapter, specifically rule 109.
(Also not new.)

maxim-tschumak replied:
Should I add a reference to `109`?

Btw, @ePaul, you are the first person who referenced to a rule by its new ID :blush:

ePaul commented on chapters/events.adoc:
```diff
@@ -0,0 +1,795 @@
+
+If a resource can be read synchronously via a REST API and read
+asynchronously via an event, the same read-permission must apply: We
+want to protect access to data, not the way data is accessed.
```
I guess we should reconsider this (not now, sometime later), now that Nakadi has a new permission model, and got rid of the permissions.

ePaul commented on chapters/http.adoc:
```diff
@@ -0,0 +1,486 @@
+*Note:* GET requests on collection resources should provide a sufficient
+filter mechanism as well as <<pagination>>.
+
+=== GET with Body
```
(Not new)
I think we should put the title in quotes ("GET with Body"), as there is no such thing in HTTP.

ePaul commented on chapters/http.adoc:
```diff
@@ -0,0 +1,486 @@
+`406 Not Acceptable` - `408 Request Timeout` - `413 Payload Too Large` -
+`414 URI Too Long` - `415 Unsupported Media Type` -
+`500 Internal Server Error` - `502 Bad Gateway` -
+`503 Service Unavailable` - `504 Gateway Timeout`
```
This list got broken. Please reformat it.

ePaul commented:
**A note about my comments:** most don't need to be fixed right now (some do, which are about formatting issues introduced by this change). But as this marks about everything as changed, I used the chance to review everything (not totally finished yet, but I've to go home now).

Feel free to fix just the formatting issues, I can open separate pull requests afterwards (or issues where there is still some need to discuss).

maxim-tschumak commented:
@ePaul Thanks for the review! - I've fixed the layout problems and the minor content improvements. (and also updated the [generated HTML](https://maxim-tschumak.github.io/restful-api-guidelines/))

@all Please add the minor, nice-to-have improvements to #283.
Reason: In order to continue working on the content of the Guidelines, we have to merge this PR as soon as possible (to avoid merge conflicts).

ePaul commented on chapters/hyper-media.adoc:
```diff
@@ -0,0 +1,150 @@
+https://www.infoq.com/news/2014/03/rest-at-odds-with-web-apis%20for%20detailed%20discussion[RESTistential
+Crisis over Hypermedia APIs],
+https://jeffknupp.com/blog/2014/06/03/why-i-hate-hateoas/[Why I Hate
+HATEOAS] and others): - We follow the <<100,API First principle>>
```
Another broken list here.

ePaul commented on chapters/json-guidelines.adoc:
```diff
@@ -0,0 +1,128 @@
+idiomatic form across teams and services.
+
+[#117]
+== {MUST} Use Consistent Property Names
```
This was previously a level-2 title above the following level-3 rules. Now it looks like it is at the same level as the rules. (Same with rule 121 below.)

maxim-tschumak replied:
This is done on purpose - the idea is to have all the rules on the same level.

ePaul commented on chapters/json-guidelines.adoc:
```diff
@@ -0,0 +1,128 @@
+[#124]
+== {SHOULD} Empty array values should not be null
+
+Empty array values can unambiguously be represented as the the empty
```
Duplicated "the".

ePaul commented on chapters/pagination.adoc:
```diff
@@ -0,0 +1,119 @@
+entries if rows are inserted or deleted between two subsequent paging
+requests.
+** When using cursor-based pagination, paging cannot continue when the
+cursor entry has been deleted while fetching two pages
```
Not new, but I highly doubt this argument. (I guess it depends on the kind of cursor you are using.)

ePaul commented on chapters/performance.adoc:
```diff
@@ -0,0 +1,167 @@
+
+Common techniques include:
+
+* gzip compression
```
this could be a link to <<156>>.

ePaul commented on chapters/performance.adoc:
```diff
@@ -0,0 +1,167 @@
+
+* gzip compression
+* querying field filters to retrieve a subset of resource attributes
+(see <<156>> below)
```
s/156/157/

tkrop replied:
@maxim-tschumak Have you checked all referenced up from the missing one? Which was it? `116`?

ePaul commented on chapters/proprietary-headers.adoc:
```diff
@@ -0,0 +1,128 @@
+Because of this, the guidelines restrict which `X-` headers can be used
+and how they are used.
+
+The Internet Engineering Task Force's states in
```
Not new, but the `'s` here is wrong. (Actually I would reword the whole paragraph, will do in a separate PR.)

ePaul commented on chapters/proprietary-headers.adoc:
```diff
@@ -0,0 +1,128 @@
+for usage so far. Remember that HTTP header field names are not
+case-sensitive.
+
+[cols="15%,10%,60%,15%",options="header",]
```
Do we really need to define the column widths manually? Isn't this something a browser can figure out based on the table content?

maxim-tschumak replied:
Unfortunately, no. The browser renders all the columns with the same width (when using the default CSS).

tkrop replied:
@maxim-tschumak Have you checked that the style uses `table-layout: auto`? This should provide what @ePaul  expects.

ePaul commented on chapters/resources.adoc:
```diff
@@ -0,0 +1,196 @@
+
+However, they also come with some disadvantages:
+
+* pure technical key without meaning; not ready for naming or name scope
```
(Not new.) This list looks like each sentence there should be start with "they are ...". Maybe put this at the end of the paragraph above?

ePaul commented on chapters/resources.adoc:
```diff
@@ -0,0 +1,196 @@
+
+However, if the resource can be accessed directly via its unique id,
+then the API should expose it as a top-level resource. For example,
+customer is a collection for sales orders; however, sales orders have
```
is → has

wssbck commented:
Looks great. One more addition perhaps - it would be very cool if headers in the navigation on the left had some highlighting for currently browsed section.

tkrop commented:
@wssbck 😸 we have this on the todo-list #283.

tfrauenstein commented:
👍  look great -- thx.

tkrop commented:
👍

ePaul commented:
@tfrauenstein do you want to change numbers before we merge it, or do we risk breaking links if you do it later?

ePaul commented:
:+1:

tkrop commented:
:+1:

maxim-tschumak commented:
:+1:
